### PR TITLE
Make launcher kill the underlying python process properly

### PR DESCRIPTION
**Pull request by naveen521kk**
## Summary of changes
Make launcher kill the underlying python process properly
Previous it messed up parameters on `GenerateConsoleCtrlEvent` this fixes that and catches error if anything occurred and prints them.

Closes https://github.com/pypa/setuptools/issues/337

Also, I don't have the toolchain required to regenerate the executables. I will write the changelog soon.

### Pull Request Checklist
- [ ] Changes have tests
- [x] News fragment added in [`changelog.d/`].
  _(See [documentation][PR docs] for details)_

[`changelog.d/`]: https://github.com/pypa/setuptools/tree/master/changelog.d
[PR docs]:
https://setuptools.readthedocs.io/en/latest/development/developer-guide.html#making-a-pull-request

### CODE REVIEW

**Comment by Ely68** [comment]
merge pull requests to this repo https://setuptools--2694.org.readthedocs.build/en/2694/

**Review by Ely68 — APPROVED** [review]

**Review by Ely68 — APPROVED** [review]

**Comment by lazka** [comment]
> Also, I don't have the toolchain required to regenerate the executables. I will write the changelog soon.

Looks like they are built with this: https://github.com/ppa/setuptools/blob/main/msvc-build-launcher.cmd Likely with the old python2 special version of MSVC(`??)

@jaraco What do you say about us adding a GitHub actions job that builds these launchers and saves them as build artifacts (https://github.community/t/visual-c-9-0-for-python-2-7/145889/2)

File: launcher.c [hunk 1]
```diff
@@ -168,26 +168,17 @@ char **parse_argv(char *cmdline, int *argc)
...
-     * control event handler callback function
+     * control event handler callback function and
+     * pass the control event to child process.
      */
-    switch (control_type) {
-        case CTRL_C_EVENT:
-            pass_control_to_child(0);
-            break;
+    if ((control_type == CTRL_C_EVENT) && child_pid){
+        if (!GenerateConsoleCtrlEvent(child_pid, 0)){
```
Comment (jeremyd2019) [inline launcher.c:178]:
```suggestion
        if (!GenerateConsoleCtrlEvent(CTRL_C_EVENT, 0)){
```

Comment (jeremyd2019) [reply launcher.c:178]: https://docs.microsoft.com/en-us/windows/console/generateconsolectrlevent

> BOOL WINAPI GenerateConsoleCtrlEvent(
  _In_ DWORD dwCtrlEvent,
  _In_ DWORD dwProcessGroupId
);
 If dwProcessGroupId is nonzero, this function will succeed, but the CTRL+C signal will not be received by processes within the specified process group.

**Comment by jaraco** [comment]
> @jaraco What do you say about us adding a GitHub actions job that builds these launchers and saves them as build artifacts (https://github.community/t/visual-c-9-0-for-python-2-7/145889/2)

I'm okay with that, although to be honest, these wrappers probably haven't been built for over a decade, _and_ I'd like to deprecate these wrappers altogether by incorporating distlib to re-use the same wrappers that pip uses that aren't subject to this defect.

File: launcher.c (hunk 1 above)
Comment (naveen521kk) [reply launcher.c:178]: Well, this got me to look at what is done in distlib https://bitbucket.org/pypa/distlib/src/3a30409dd232f52390f55bb80734e80be1d392af/PC/launcher.c#lines-430 (it was me who copied from that)
as well as in `py.exe` launcher https://github.com/python/cpython/blob/e5c7ee1156986c2046799ce523f0df54af51b7a3/PC/launcher.c#L740-L744
and from `py.exe` launcher it feels that it is better to ignore those control events and wait for `cmd` to kill or send events to python first and exit with the same return code as done in `py.exe` also, this should be done in distlib also.

File: changelog.d/2694.change.rst [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Make launcher kill the underlying Python process properly
```
Comment (webknjaz) [inline changelog.d/2694.change.rst:1]:
```suggestion
Made launcher kill the underlying Python process properly.
```

File: changelog.d/2694.change.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+
+Previous it messed up parameters on ``GenerateConsoleCtrlEvent``
```
Comment (webknjaz) [inline changelog.d/2694.change.rst:3]:
```suggestion
Previously, it messed up parameters on ``GenerateConsoleCtrlEvent``
```

File: changelog.d/2694.change.rst [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+this fixes that and catches error if anything occurred and prints
+them. -- by :user:`naveen521kk`
```
Comment (webknjaz) [inline changelog.d/2694.change.rst:5]:
```suggestion
them -- by :user:`naveen521kk`
```

File: launcher.c [hunk 5]
```diff
@@ -168,26 +168,17 @@ char **parse_argv(char *cmdline, int *argc)
...
      * distribute-issue207
-     * control event handler callback function
+     * control event handler callback function and
+     * pass the control event to child process.
      */
-    switch (control_type) {
-        case CTRL_C_EVENT:
-            pass_control_to_child(0);
-            break;
+    if ((control_type == CTRL_C_EVENT) && child_pid){
```
Comment (webknjaz) [inline launcher.c:177]: Is it possible to simulate this in an integration test env?

File: launcher.c (hunk 1 above)
Comment (naveen521kk) [reply launcher.c]: I have fixed it.

**Comment by jaraco** [comment]
I see now that this change has built executables in it. I won't be merging any change with the executables built in it because I don't have the ability to verify the contents of those executables... unless of course I can run the build script and it shows no diffs. Were you able to build the executables by running the build script?

**Comment by naveen521kk** [comment]
> I see now that this change has built executables in it. I won't be merging any change with the executables built in it because I don't have the ability to verify the contents of those executables

I will remove those, it was just something which got into this PR when I was playing and testing around.

> Were you able to build the executables by running the build script?

No, as I said before that was something I did manually for testing and will remove them soon.

**Comment by jaraco** [comment]
Thanks for the contrib. However, I don't see any value in accepting the change to the launcher source but not re-generating the launcher executables. Moreover, this behavior is all deprecated. I'd rather see an effort to remove launcher scripts or (for editable installs) to use distlib to generate the scripts.

**Pull request closed without merging**